### ISSUE TITLE
Cross compilation runtime builds require compiler tools

### ISSUE THREAD

**Issue by gmmyung** [iree-org/iree#16109]
### Request description

Currently, compiler tools such as `iree-compile` and `iree-tblgen` is required when building with `IREE_HOST_BIN_DIR` enabled. This forces user who cross compiles the runtime set `IREE_BUILD_COMPILER=ON` on their host build.

https://github.com/openxla/iree/blob/914f30641f3c41338f8a7820711743ca49a2a76f/tools/CMakeLists.txt#L35C1-L46C1

It will be a lot easier if the runtime cross compilation does not require the compiler tools.

### What component(s) does this issue relate to?

Other

**Comment by ScottTodd** [comment]
Can you elaborate on what configuration you want to work? (Full CMake configure commands for host and target would be ideal)

One idea: this check could be changed to a warning and an early return: https://github.com/openxla/iree/blob/914f30641f3c41338f8a7820711743ca49a2a76f/build_tools/cmake/iree_import_binary.cmake#L60-L64

A few parts of the build could fail later during configuration, depending on which other options are set, like `iree_check_test` when `IREE_BUILD_TEST` is enabled (that needs `iree-compile`). If there are only a few places then the warnings/errors could just be pushed further down into other CMake functions.

Actually, I have https://github.com/openxla/iree/pull/16086 open and that could check for the presence of `iree-compile` here instead:
```cmake
  # 1. Check for the compiler.
  if(NOT IREE_BUILD_COMPILER AND NOT IREE_HOST_BIN_DIR)
    set(_BYTECODE_MODULE_BUILD_ENABLED FALSE)
  endif()
```

**Comment by gmmyung** [comment]
I am working on embedded target (thumb, riscv) support for a Rust binding. I am using [runtime-library](https://github.com/iree-org/iree-samples/tree/main/runtime-library) for the target build.
Build config for host installation
```rust
cmake::Config::new(iree_path)
    .define("BUILD_SHARED_LIBS", "OFF")
    .define("IREERT_ENABLE_LTO", "OFF")
    .define(
        "IREE_ROOT_DIR",
        PathBuf::from(env!("CARGO_MANIFEST_DIR"))
            .join("iree")
            .to_str()
            .unwrap(),
        )
    .define(
        "CMAKE_INSTALL_PREFIX",
        build_path.join("install").to_str().unwrap(),
    )
    .define("CMAKE_C_COMPILER", "clang")
    .define("CMAKE_CXX_COMPILER", "clang++")
    .define("IREE_HAL_DRIVER_DEFAULTS", "OFF")
    .define("IREE_HAL_DRIVER_LOCAL_SYNC", "ON")
    .define("IREE_BUILD_COMPILER", "ON")
    .define("IREE_BUILD_TESTS", "OFF")
    .define("IREE_BUILD_SAMPLES", "OFF")
    .build_target("install")
    .target(&std::env::var("HOST").unwrap())
    .build();
```
Config for target install
```rust
cmake::Config::new(runtime_lib_path)
    .define("BUILD_SHARED_LIBS", "OFF")
    .define("IREERT_ENABLE_LTO", "OFF") // TODO: Enable LTO
    .define(
        "IREE_ROOT_DIR",
        PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("iree")
        .to_str()
        .unwrap(),
    )
    .define("IREE_BUILD_COMPILER", "OFF");
    .define("CMAKE_SYSTEM_NAME", "Generic")
    .define("CMAKE_EXE_LINKER_FLAGS_INIT", "--specs=nosys.specs -lc -lm")
    .define("IREE_HOST_BIN_DIR", build_path.join("install").join("bin"))
    .cflag("-DIREE_PLATFORM_GENERIC=1")
    .cflag("-Wno-char-subscripts")
    .cflag("-Wno-format")
    .cflag("-Wno-implicit-function-declaration")
    .cflag("-Wno-unused-variable")
    .cflag(r#"-DIREE_TIME_NOW_FN="{ return 0; }" -DIREE_WAIT_UNTIL_FN=wait_until"#)
    .define("IREE_ENABLE_THREADING", "OFF")
    .define("IREE_HAL_DRIVER_DEFAULTS", "OFF")
    .define("IREE_HAL_DRIVER_LOCAL_SYNC", "ON")
    .define("IREE_HAL_EXECUTABLE_LOADER_DEFAULTS", "OFF")
    .define("IREE_HAL_EXECUTABLE_LOADER_EMBEDDED_ELF", "ON")
    .define("IREE_HAL_EXECUTABLE_LOADER_VMVX_MODULE", "ON")
    .define("IREE_HAL_EXECUTABLE_PLUGIN_DEFAULTS", "OFF")
    .define("IREE_HAL_EXECUTABLE_PLUGIN_EMBEDDED_ELF", "ON")
    .build();
```
It would be helpful if cross-compiled runtime builds can be set in a way that does not require the iree compiler tools, since Rust packages can be no bigger than 10MB. I was able to cram the IREE runtime build inside my Rust package for native builds, but the IREE compiler build is way too large to pack inside a Rust package.

**Comment by stellaraccident** [comment]
There's definitely something wrong here: we build the runtime library standalone all the time, so I know that is well supported. But we need to unpack what you are trying with respect to cross compilation, etc.

Forgive me for not being fully up to speed on cargo and how you are using it. Is the above using iree libraries and then cargo is just referencing the headers/built libs (vs trying to build the runtime sources with cargo)?

We are making progress on graduating that runtime library sample into the main project. Ideally, we have someone make sure that the cross compilation flow works and is documented with the dev library install path: https://discord.com/channels/689900678990135345/1083185300546793553/1190447652421243021

(I expect it should work today but we should work out the exact flags)

**Comment by stellaraccident** [comment]
Do you happen to have a repo or something you could share? I'm interested in upping my rust/cargo game and helping but don't want to just blindly reproduce what you've done.

**Comment by stellaraccident** [comment]
Ok, #16111 gets this working for me. Here, I downloaded the arm-none-eabi toolchain from ARM's website and combined various sources until it built. This just uses a stock IREE checkout with just the runtime submodules as per `./build_tools/scripts/git/update_runtime_submodules.sh`. The sample toolchain repo is no longer needed as what it did is all upstream now.

```
# the name of the target operating system
set(CMAKE_SYSTEM_NAME Generic)

set(_ARM_TOOLCHAIN_ROOT /Applications/ArmGNUToolchain/13.2.Rel1/arm-none-eabi)

# Disable some linkages not supported on bare metal.
set(CMAKE_EXE_LINKER_FLAGS_INIT "--specs=nosys.specs")

# which compilers to use for C and C++
set(CMAKE_C_COMPILER   ${_ARM_TOOLCHAIN_ROOT}/bin/arm-none-eabi-gcc)
set(CMAKE_CXX_COMPILER ${_ARM_TOOLCHAIN_ROOT}/bin/arm-none-eabi-g++)

# where is the target environment located
set(CMAKE_FIND_ROOT_PATH  ${_ARM_TOOLCHAIN_ROOT})

# adjust the default behavior of the FIND_XXX() commands:
# search programs in the host environment
set(CMAKE_FIND_ROOT_PATH_MODE_PROGRAM NEVER)

# search headers and libraries in the target environment
set(CMAKE_FIND_ROOT_PATH_MODE_LIBRARY ONLY)
set(CMAKE_FIND_ROOT_PATH_MODE_INCLUDE ONLY)

# Some IREE flags for bare metal.
set(_IREE_C_FLAGS)
set(_IREE_C_FLAGS "-mcpu=cortex-m4 -mthumb -mno-thumb-interwork -mfloat-abi=softfp ")
string(APPEND _IREE_C_FLAGS "-DIREE_PLATFORM_GENERIC=1 ")
string(APPEND _IREE_C_FLAGS "-DIREE_FILE_IO_ENABLE=0 ")
string(APPEND _IREE_C_FLAGS "-DIREE_TIME_NOW_FN=\"\{ return 0;\}\" ")
string(APPEND _IREE_C_FLAGS "-D'IREE_WAIT_UNTIL_FN(n)=\"{ return false; }\"' ")
string(APPEND _IREE_C_FLAGS "-DIREE_SYNCHRONIZATION_DISABLE_UNSAFE=1 ")
string(APPEND _IREE_C_FLAGS "-DFLATCC_USE_GENERIC_ALIGNED_ALLOC=1 ")
string(APPEND _IREE_C_FLAGS "-DIREE_STATUS_FEATURES=0 ")
string(APPEND _IREE_C_FLAGS "-Wno-char-subscripts ")
string(APPEND _IREE_C_FLAGS "-Wno-format ")
string(APPEND _IREE_C_FLAGS "-Wno-unused-variable ")
string(APPEND _IREE_C_FLAGS "-Wl,--gc-sections -ffunction-sections -fdata-sections")

set(CMAKE_C_FLAGS "${_IREE_C_FLAGS} -Wno-implicit-function-declaration ${CMAKE_C_FLAGS}")
set(CMAKE_CXX_FLAGS "${_IREE_C_FLAGS} ${CMAKE_CXX_FLAGS}")

unset(_IREE_C_FLAGS)
unset(_ARM_TOOLCHAIN_ROOT)
```

This command to build the host runtime/tools:

```
cmake -GNinja -S. -Bbuild \
  -DIREE_BUILD_COMPILER=OFF -DIREE_HAL_DRIVER_DEFAULTS=OFF \
  -DIREE_BUILD_TESTS=OFF -DIREE_BUILD_SAMPLES=OFF
cmake --build build
```

This command to build the target runtime:

```
cmake -GNinja -S. -Bbuild-cross/Generic \
  -DCMAKE_TOOLCHAIN_FILE=$HOME/bin/arm-none-eabi-toolchain.cmake \
  -DIREE_HOST_BIN_DIR=$PWD/build/tools \
  -DCMAKE_INSTALL_PREFIX=build-cross/Generic/install
  -DIREE_BUILD_TESTS=OFF -DIREE_BUILD_SAMPLES=OFF \
  -DIREE_BUILD_COMPILER=OFF -DIREE_ENABLE_THREADING=OFF -DIREE_HAL_DRIVER_DEFAULTS=OFF \
  -DIREE_HAL_DRIVER_LOCAL_SYNC=ON -DIREE_HAL_EXECUTABLE_LOADER_DEFAULTS=OFF \
  -DIREE_HAL_EXECUTABLE_LOADER_EMBEDDED_ELF=ON \
  -DIREE_HAL_EXECUTABLE_LOADER_VMVX_MODULE=ON \
  -DIREE_HAL_EXECUTABLE_PLUGIN_DEFAULTS=OFF \
  -DIREE_HAL_EXECUTABLE_PLUGIN_EMBEDDED_ELF=ON
cmake --build build-cross/Generic
cmake --build build-cross/Generic --target iree-install-dist
```

If it worked entirely, you should find headers in `build-cross/Generic/install/include` and a unified runtime library in `build-cross/Generic/install/lib/libiree_runtime_unfied.a`.

I doubt this will work in any real system without more arch tweaks. However, building is a first step. If doing this for real, I would have defined all of the arch tweaks and needed stubs in a header file and included it with `-include` in the CFLAGS. That way, you could provide real time functions and work around missing platform goo (for some reason ARM's toolchain doesn't define posix_memalign and we just carefully skate around it above).

**Comment by stellaraccident** [comment]
Landed the necessary patch and the above (revised) instructions worked for me at head (where "worked" == "built", I have no hardware to test on).

**Comment by gmmyung** [comment]
> Do you happen to have a repo or something you could share? I'm interested in upping my rust/cargo game and helping but don't want to just blindly reproduce what you've done.

https://github.com/gmmyung/eerie/blob/embedded/iree-sys/build.rs

Cargo compiles and runs `build.rs` on the host machine prior to compiling the actual library, so we can run extra steps such as invoking cmake, auto-generate rust bindings from the c header files, and many more.

**Comment by gmmyung** [comment]
I will update my dependency to `main` and check if things work out for me.

**Comment by gmmyung** [comment]
I have reached this error, is this related to #16110?
```
make: warning: jobserver unavailable: using -j1.  Add `+' to parent make rule.
  In file included from /Users/gmmyung/Developer/eerie/iree-sys/iree/runtime/src/iree/hal/driver_registry.c:12:
  /Users/gmmyung/Developer/eerie/iree-sys/iree/runtime/src/iree/base/internal/call_once.h:90:29: error: 'PTHREAD_ONCE_INIT' undeclared here (not in a function)
     90 | #define IREE_ONCE_FLAG_INIT PTHREAD_ONCE_INIT
        |                             ^~~~~~~~~~~~~~~~~
  /Users/gmmyung/Developer/eerie/iree-sys/iree/runtime/src/iree/hal/driver_registry.c:44:5: note: in expansion of macro 'IREE_ONCE_FLAG_INIT'
     44 |     IREE_ONCE_FLAG_INIT;
        |     ^~~~~~~~~~~~~~~~~~~
  make[2]: *** [iree_core/runtime/src/iree/hal/CMakeFiles/iree_hal_hal.objects.dir/driver_registry.c.obj] Error 1
  make[1]: *** [iree_core/runtime/src/iree/hal/CMakeFiles/iree_hal_hal.objects.dir/all] Error 2
  make: *** [all] Error 2
  thread 'main' panicked at /Users/gmmyung/.cargo/registry/src/index.crates.io-6f17d22bba15001f/cmake-0.1.50/src/lib.rs:1098:5:
```

**Comment by stellaraccident** [comment]
> string(APPEND _IREE_C_FLAGS "-DIREE_SYNCHRONIZATION_DISABLE_UNSAFE=1 ")

I did hit that. I can't remember which of the grab bag of things got past it. I think this one:

```
string(APPEND _IREE_C_FLAGS "-DIREE_SYNCHRONIZATION_DISABLE_UNSAFE=1 ")
```

(we should treat each of those carve-outs as bugs and make the system auto-detct, but one step at a time)

**Comment by benvanik** [comment]
IREE_SYNCHRONIZATION_DISABLE_UNSAFE is only valid to use on single core systems with no threading - it's hard to autodetect as even without pthreads/etc you can have multiple threads and need synchronization. It's the most dangerous flag we have so autodetection is risky (too risky for me to support).

**Comment by benvanik** [comment]
(if there was a platform flag already set we could use to infer that'd be ok - but just turning all atomic operations/mutices/etc into no-ops is real risky to do with magic)

**Comment by stellaraccident** [comment]
Right. But we could still have a check that basic synchronization features exist on the platform and provide a more verbose error/suggestion.

**Comment by benvanik** [comment]
For sure warning sounds good. The tricky part of all this is that on bare metal you (today) need to bring your own impl as even single core systems with no OS (or a minimal RTOS) can still have interrupt handlers and such (for user code launching/sequencing calls into us or for HAL backends that use DSPs/accelerators, etc). Warnings feel like a good way to make the compilation errors in the default case make a bit more sense and also let a user know that part of doing bare metal development is setting esoteric/sketchy flags :P

We could break this flag up to allow more to work automatically and reduce the risk of flipping it. The major things controlled by the flag are atomics (which we could make a compiler setting as to whether clang/gcc/msvc have them instead of relying on the flag), call_once (which we could implement with atomics ourselves and piggy-back on the atomics), and full mutex/futex-like things like wait handles and notifications - those are really the only ones that get pretty platform-specific. Would need to get a few big targets we want to claim support for together then could figure out how to slice things/how much we can just make it work.

Many embedded systems will still have synchronization even if they don't have threads (like mbed's bare metal profile which is popular with platformio https://os.mbed.com/docs/mbed-os/v6.16/bare-metal/index.html#features, but most RTOS or mini-RTOS-like platforms have similar configs). I can't think of many cases where it's worth the risk to disable synchronization for both user code interacting with IREE and for IREE internals instead of just implementing it for the platform (it doesn't save much code size or runtime performance given how coarse/infrequently we synchronize). So I think if we made most of it work everywhere (reasonably) we may be able to reduce the times that users run into this (but still would be nice to have a warning for those that do).

**Comment by stellaraccident** [comment]
It's been a number of years since I've done real bare metal or RTOS development, but I think this might be the job of having board support files vs a bunch of loose options that work around sketchy things in a way that is very easy to copypasta.

A template board support header with commented out defines, descriptions and some preprocessor based assertions would be more robust.

Smaller systems often have different cores with different capabilities, and it is totally reasonable for some of them to be "very" bare metal. I think it just needs to be done in a way that makes it clearly the user's decision. If I were using an RTOS with preemption enabled, I should be very sketched out by some of these options, especially if there is a support header next to it for my RTOS which clearly enables the preemption safeties.

Anyway... All for later.

**Comment by benvanik** [comment]
Agreed; we have `IREE_USER_CONFIG_H` we can use for some of that (`-DIREE_USER_CONFIG_H=my_platform_config.h`).

**Comment by gmmyung** [comment]
By the way, I am having a really bad time linking Newlib/libc functions to my Rust binary:(
Are there any bare-metal feature gates that disables compilation of file, thread, alloc, assert related libc calls?

**Comment by gmmyung** [comment]
Also, does this mean that IREE constructs are not thread-safe on embedded environments with `IREE_SYNCHRONIZATION_DISABLE_UNSAFE` set? 

> IREE_SYNCHRONIZATION_DISABLE_UNSAFE is only valid to use on single core systems with no threading - it's hard to autodetect as even without pthreads/etc you can have multiple threads and need synchronization. It's the most dangerous flag we have so autodetection is risky (too risky for me to support).

**Comment by benvanik** [comment]
Correct - if you set the magic flag that unsafely disables all synchronization then it's not thread safe :)

You can check iree/base/config.h for most settings - there's the C macro IREE_FILE_IO_ENABLE that can be used to disable file I/O (though only our tooling/flags/etc really uses that, and it's not bare-metal portable). The cmake IREE_ENABLE_THREADING that can be used to disable threading. Note that you lose the functionality guarded by these flags - if you want multi-threaded CPU execution then you need threads, etc.

There is no way today to remove _all_ libc calls - you'll have to provide your own either via a libc-like (musl/etc) or your own definitions for what's used (that may then just wrap and route to whatever else you want), but in minimal configurations that mostly just means math stuff and the normal helpers (memcpy/memmove/etc). Assert can be avoided with things like NDEBUG (asserts are debug only), but you're expected to have a basic C environment (approximation) today. Note that some things like malloc are only used by a few features - outside of runtime/src/iree/tooling/ (which is not bare-metal portable) the only place that uses iree_allocator_system and pulls in malloc is the status handling, which if you disable status messages (IREE_STATUS_MODE=0) then shouldn't get limited in (I really don't recommend doing that, though, as you're going to have a hard time without status messages).

**Comment by stellaraccident** [comment]
> Also, does this mean that IREE constructs are not thread-safe on embedded environments with `IREE_SYNCHRONIZATION_DISABLE_UNSAFE` set?
> 
> > IREE_SYNCHRONIZATION_DISABLE_UNSAFE is only valid to use on single core systems with no threading - it's hard to autodetect as even without pthreads/etc you can have multiple threads and need synchronization. It's the most dangerous flag we have so autodetection is risky (too risky for me to support).

There is a big difference between embedded and bare metal. If there is an OS at all which supports preemption, you really need to not use this flag and make sure to map the synchronization support properly.

On true bare metal, it is just you, the processor and your interrupt routines. Those are the cases where you will want to use such a flag and manage it yourself in some way.

True bare metal has a lot of details that need to line up. Newlib adds some very primitive support but it's not the only thing out there. I'd need to know more about what you are trying to do and what assumptions rust is making. It is not surprising to me that these things are hard to wedge together without knowing exactly what is going on.

**Comment by stellaraccident** [comment]
I assume you are trying to build a no_std crate... Ok. So what were actually looking for is a way to build with no dep on newlib at all. As Ben says, it's likely not too far off but will need some things from libm, etc. it is also possible/likely that by building the whole project, you're getting tooling and libraries that use newlib stuff more -- but that isn't used in the actual runtime flow.

If I were doing this, I would:

1. Use all of the flags Ben mentions to disable features.
2. Possibly build a stripped down set of targets.
3. Use a user config header to provide inline stubs for any leftovers you want to buy be served by newlib.

**Comment by stellaraccident** [comment]
Unlike rust itself, there are some unavoidable deps on the "system" which need to be resolved somehow. Either via newlib or bringing your own.

**Comment by gmmyung** [comment]
Yes, I am trying to use [tinyrlibc](https://github.com/rust-embedded-community/tinyrlibc) to provide substitutes for libc functions. A few more questions:
- Does IREE use errno flags extensively, or only for status messages? (I have to provide a rust errno implementation if so)

> Note that some things like malloc are only used by a few features - outside of runtime/src/iree/tooling/ (which is not bare-metal portable) the only place that uses iree_allocator_system and pulls in malloc is the status handling.

- Does malloc used in status handling have alignment requirements?

**Comment by benvanik** [comment]
we don't use errno (besides when interacting with parts of the system that use errno, which in minimal builds aren't usually included)

yes the status alloc requires alignment, but if posix_memalign is not available we support that ourselves (see iree_aligned_alloc in iree/base/status.c)
